File: C/linked.c
#include <stdio.h>

typedef struct _Node
{
    int data;
    Node *next;
} Node;

Node *reverse(Node *head)
{
    if (head == NULL || head->next == NULL)
        return head;

    Node *next = NULL;
    Node *prev = NULL;

    while (head != NULL)
    {
        next = head->next;
        head->next = prev;
        prev = head;
        head = next;
    }
    return prev;
}

Node *middle(Node *head)
{
    if (head == NULL || head->next == NULL)
        return head;

    Node *fast = head;
    Node *low = head;
    while (head != NULL && head->next != NULL)
    {
        fast = fast->next->next;
        low = low->next;
    }
    return low;
}

int is_cycled(Node *head)
{
    if (head == NULL || head->next == NULL)
        return 0;

    Node *fast = head;
    Node *low = head;
    while (head != NULL && head->next != NULL)
    {
        fast = fast->next->next;
        low = low->next;

        if (fast == low)
        {
            return 1;
        }
    }
    return 0;
}

File: Java/src/linked/Linked.java
package linked;

import base.Node;

public class Linked {

    /**
     * Reverse the linked list.
     */
    public static Node reverse(Node list) {
        Node head = null;
        Node prev = null;
        Node cur = list;
        while (cur != null) {
            Node next = cur.next;
            if (next == null) {
                head = cur;
            }

            cur.next = prev;
            prev = cur;
            cur = next;
        }

        return head;
    }

    /**
     * Whether the linked list is cycled.
     */
    public static boolean isCycled(Node list) {
        if (list == null) {
            return false;
        }

        Node slow = list;
        Node fast = list.next;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                return true;
            }
        }

        return false;
    }

    /**
     * Find the middle node in linked list.
     */
    public static Node findMiddle(Node list) {
        if (list == null) {
            return null;
        }

        Node slow = list;
        Node fast = list;
        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }

    /**
     * Whether the char linked list is a palindrome.
     */
    public static boolean isPalindrome(Node list) {
        return false;
    }
}